### ISSUE TITLE
Statistical report for constant time compare tests?

### ISSUE THREAD

**Issue by ericelliott** [ncb000gt/node.bcrypt.js#266]
Hi,

I'm the maintainer of Credential. We're currently trying to harden our constant time equality check in Credential, and we're looking for a good statistics test to ensure that our timing is constant enough to thwart timing attacks.

Your experience and feedback would be very useful. Please take a look at this [proposed constant time string comparison](https://github.com/ericelliott/credential/pull/13). I'm especially interested in producing a test suite that can guarantee statistically significant constant time equality comparisons, and I believe such tests would be useful to node.bcrypt, as well.

Thanks!

**Comment by defunctzombie** [comment]
You should check out this module: https://www.npmjs.com/package/scmp

Otherwise, posting on stack overflow will be better, not here.

**Comment by ericelliott** [comment]
My assertion is that none of our libs are complete without a statistical proof that our constant time comparisons are constant enough to thwart actual timing attacks. See Crosby et al: 

> "We have shown that, even though the Internet induces significant timing jitter, we can reliably distinguish remote timing differences as low as 20 µs. A LAN environment has lower timing jitter, allowing us to reliably distinguish remote timing differences as small as 100 ns (possibly even smaller). These precise timing differences can be distinguished with only hundreds or possibly thousands of measurements." Crosby et al., ["Opportunities and Limits of Remote Timing Attacks"](http://www.cs.rice.edu/~dwallach/pub/crosby-timing2009.pdf)

So if we're going to implement libraries and call them secure, we need to prove it in our test suites. Wouldn't you agree? With Credential, I don't just expect people to trust that it's secure. [I'm putting my money where my mouth is.](https://github.com/ericelliott/credential/issues/8).

Because what we do in our libraries potentially impacts the personal security of millions of people, I think it's our responsibility as security library maintainers to provide assurances that we take these things seriously.

**Comment by ncb000gt** [comment]
Strictly speaking, hash functions like bcrypt and scrypt aren't "vulnerable" to timing attacks. I may update the comparison function at some point, but a timing attack would only reveal information about the hash itself and not the values inside since the comparison tests one hash against another. Does that make sense?

Take a look at http://security.stackexchange.com/a/46215 for more, and better IMO, information.

**Comment by ericelliott** [comment]
That comment is the result of a complete misunderstanding about how timing attacks work. Timing attacks don't have to guess a correct password to achieve a breach. They only need to find a string that produces the same hash. In other words, even finding a hash collision (not the user's password) can give an attacker access to the account in question, and if they find a colliding hash, they have a relevant chance that the string they found *is* the user's password, which matters not at all for that one system, but does have the implication that the attacker is one step closer to gaining access to that user's other accounts on other systems.

Timing attacks work in a similar manner to how playing hangman works, but they're not about guessing the user's password. They're about finding a string that hashes to the same string as the stored hash -- which is *easier than finding the real password.*

In other words, timing attacks are about tricking the system into letting you into a particular user's account, regardless of whether or not you've actually found the password.

It's really very simple: If you can find a string that results in the password hash library returning *true,* you can use that string to get into the user's account, even if that string is not the user's real password.

**Comment by jfirebaugh** [comment]
> They only need to find a string that produces the same hash.

That's called a first-preimage attack. It would be astonishing if bcrypt was vulnerable to first-preimage attacks, and in such an eventuality, the only responsible action would be to abandon it entirely; a constant-time hash comparison algorithm would be worthless. And due to the nature of how one-way hashes work, a timing attack does not aid in _finding_ a preimage -- that's where you may be getting confused.

For more discussion of this topic, see https://github.com/codahale/bcrypt-ruby/pull/43.

**Comment by ericelliott** [comment]
Thanks @jfirebaugh. I'm not a security professional, but I'm very interested in getting this right: not just getting it right, but proving that we've thought these considerations through and put all our assumptions to the test.

If it's worth putting constant time equality into our libraries, then it's also worth proving that they're constant. Otherwise, we should use `===` and be done with it in order to reduce the complexity (read: potential attack surface) of the libraries.

If it's not worth using them, then we should be able to easily prove that, too. IMO, the best proof is a user-reproducible brute force vs timing attack demo in the repository, or linked from the library documentation.

**Comment by ericelliott** [comment]
I really appreciate all the great information in this thread. Thanks a lot. :)
